src/error_maps.jl

```diff
@@ -0,0 +1,314 @@
+struct ExponentialMap <: ErrorMap end
+struct MRPMap <: ErrorMap end
+struct QuatVecMap <: ErrorMap end
+struct IdentityMap <: ErrorMap end
+
+"""
+    InvErrorMap
+
+The nonlinear mapping from unit quaternions to a three-dimensional error state. Obtained by
+inverting an `ErrorMap`, i.e.
+
+    InvCayleyMap() = inv(CayleyMap())
+
+# Usage
+    imap(R::Rotation)             # "inverse" map from a rotation to a 3D error
+    inv(imap)(v::AbstractVector)  # "forward" map from a 3D error to a `UnitQuaternion`
+
+where `imap <: InvErrorMap`.
+
+See `ErrorMap` for documentation on the implemented maps.
+"""
+abstract type InvErrorMap end
+struct InvCayleyMap <: InvErrorMap end
+struct InvExponentialMap <: InvErrorMap end
+struct InvMRPMap <: InvErrorMap end
+struct InvQuatVecMap <: InvErrorMap end
+
+inv(::CayleyMap) = InvCayleyMap()
+inv(::ExponentialMap) = InvExponentialMap()
+inv(::QuatVecMap) = InvQuatVecMap()
+inv(::MRPMap) = InvMRPMap()
+inv(::IdentityMap) = IdentityMap()
+
+inv(::InvCayleyMap) = CayleyMap()
+inv(::InvExponentialMap) = ExponentialMap()
+inv(::InvQuatVecMap) = QuatVecMap()
+inv(::InvMRPMap) = MRPMap()
+
+
+# Scalings
+@inline scaling(::Type{ExponentialMap}) = 0.5
+@inline scaling(::Type{QuatVecMap}) = 1.0
+@inline scaling(::Type{CayleyMap}) = 1.0
+@inline scaling(::Type{MRPMap}) = 2.0
+scaling(m::M) where M <: ErrorMap = scaling(M)
+
+# Quaternion Maps
+(::ExponentialMap)(ϕ::AbstractVector) = expm(ϕ/scaling(ExponentialMap))
+
+function (::QuatVecMap)(v::AbstractVector)
+    check_length(v, 3)
+    μ = 1/scaling(QuatVecMap)
+    UnitQuaternion(sqrt(1-μ^2*v'v), μ*v[1], μ*v[2], μ*v[3])
+end
+
+function (::CayleyMap)(g::AbstractVector)
+    check_length(g, 3)
+    g /= scaling(CayleyMap)
+    M = 1/sqrt(1+g'g)
+    UnitQuaternion(M, M*g[1], M*g[2], M*g[3])
+end
+
+function (::MRPMap)(p::AbstractVector)
+    check_length(p, 3)
+    p /= scaling(MRPMap)
+    n2 = p'p
+    M = 2/(1+n2)
+    UnitQuaternion((1-n2)/(1+n2), M*p[1], M*p[2], M*p[3])
+end
+
+function (::IdentityMap)(q::AbstractVector)
+    check_length(q, 3)
+    UnitQuaternion(q[1], q[2], q[3], q[4])
+end
+
+
+# Quaternion Map Jacobians
+"""
+    jacobian(::ErrorMap, ϕ)
+
+Jacobian of the quaternion map that takes a three-dimensional vector `ϕ` and returns a
+    unit quaternion.
+Returns a 4x3 Static Matrix
+
+For all the maps (except the `IdentityMap`)
+`jacobian(::ErrorMap, zeros(3)) = [0; I] = Hmat()'`
+"""
+function jacobian(::ExponentialMap, ϕ, eps=1e-5)
+    μ = 1/scaling(ExponentialMap)
+    θ = norm(ϕ)
+    cθ = cos(μ*θ/2)
+    sincθ = sinc(μ*θ/2π)
+    if θ < eps
+        0.5*μ*[-0.5*μ*sincθ*ϕ'; sincθ*I + (cθ - sincθ)*ϕ*ϕ']
+    else
+        0.5*μ*[-0.5*μ*sincθ*ϕ'; sincθ*I + (cθ - sincθ)*ϕ*ϕ'/(ϕ'ϕ)]
+    end
+end
+
+function jacobian(::QuatVecMap, v)
+    μ = 1/scaling(QuatVecMap)
+    μ2 = μ*μ
+    M = -μ2/sqrt(1-μ2*v'v)
+    @SMatrix [v[1]*M v[2]*M v[3]*M;
+              μ 0 0;
+              0 μ 0;
+              0 0 μ]
+end
+
+function jacobian(::CayleyMap, g)
```

Comment:
Eventually (not in this PR!) we should integrate all the hand written derivatives here with the interface in ChainRulesCore.jl.

That will allow anyone using the AD tools which will build on ChainRulesCore to benefit from your derivatives without users needing to know anything about these specialized jacobians.

CC @oxinabox

Reply:
Correct. Hit me up if you have any questions